### ISSUE TITLE
Function prototype matching @fn command, but link between both is not made

### ISSUE THREAD

**Issue by vincentdupaquis** [doxygen/doxygen#9005]
Hello,

I've been meeting an issue where I have a function which has a first set of doxygen comments in the same file, and a set of other ones in another file, using the @fn command to make the link between both.

I have attached a sample .h file showing the issue, function2 & function3 have the detail field recognized, while function1 does not.

The problem is linked to the fact that although both parameters between the function declaration & the @fn command are identical, in the failing case a parameter was declared "char* p1", and doxygen seems to memorize this parameter as "char *p1" (the space is not placed identically).

Best regards,
Vincent.

[issue.zip](https://github.com/doxygen/doxygen/files/7820986/issue.zip)
.

**Comment by albert-github** [comment]
From the provided Doxyfile I guess you use doxygen version 1.8.17 (please always mention the used version).

I tried it with the current version (1.9.3 (c0b9eafbfb53286ce31e75e2b6c976ee4d345473)) and the current master (1.9.4 (dc2950946033e9dab125e389af6eafc3d1f73eac)) as well and see the problem here too.

**Comment by albert-github** [comment]
Interesting to see that when using `doxygen -d findmembers` we see in the output:
```
findMemberDocumentation(): root->type='' root->inside='' root->name='void function1' root->args='(char <em>p1, char</em> p2)' section=8000000 root->spec=0 root->mGrpId=-1
findMember(root=000002B878C1F140,funcDecl='void function1(char <em>p1, char</em> p2)',related='',overload=0,isFunc=1 mGrpId=-1 #tArgList=0 spec=0 lang=200
findMember() Parse results:
  namespaceName=''
  className=``
  funcType='void'
  funcSpec=''
  funcName='function1'
  funcArgs='(char< em >p1, char</em > p2)'
```
instead of the expected:
```
findMemberDocumentation(): root->type=`' root->inside=`' root->name=`void function1' root->args=`(char *p1, char* p2)' section=8000000 root->spec=0 root->mGrpId=-1
findMember(root=000001C98DAF0090,funcDecl=`void function1(char *p1, char* p2)',related=`',overload=0,isFunc=1 mGrpId=-1 tArgList=0000000000000000 (#=0) spec=0 lang=200
findMember() Parse results:
  namespaceName=`'
  className=``
  funcType=`void'
  funcSpec=`'
  funcName=`function1'
  funcArgs=`(char *p1, char *p2)'
  funcTempList=`'
  funcDecl=`void function1'
  related=`'
  exceptions=`'
  isRelated=0
  isMemberOf=0
  isFriend=0
  isFunc=1
```
The difference is introduced in doxygen version 1.8.15 (the version 1.8.14 was still OK).

Further observations learn us that the problem is caused in the markdown processing.
When replacing (as of version 1.9.3!) the line:
```
@fn void function1(char *p1, char* p2)
```
by
```
@fn function1(char *p1, char* p2)
```

@doxygen
- This looks like a problem where also the second argument should not be processed by the Markdown processing or something has to be done because of the emphasis replacement.
- any idea how to solve?

**Comment by vincentdupaquis** [comment]
This issue is particularly frustrating, it took me one full day to understand (by which miracle ?) what was the root cause.
Why is there this mis-placed space, I do not know, but it is a pain for the one who meets the issue :(

**Comment by albert-github** [comment]
Dove a bit deeper in it.
The documentation of the `@fn` states:
> \fn (function declaration)

where the `(...)` means till the end of the line.

Looks like the solution might be in:
```
 static const auto endOfFunc = [](const char *data_,int offset_,int size_) -> int
```
where it should be checked whether or not we reached the end of the line and no open bracket types (possibly also `<` ... `>` pairs).

Note that the function declaration of `@fn` can span multiple lines. like in:
```
/**
 * @brief My function 1
 *
 * @param p1 Parameter 1
 * @param p2 Parameter 2
 * @param p3 Parameter 3
 * @param p4 Parameter 4
 */
extern void function1(char p1, char p2, char p3, char p4);


/**
 * @fn void function1(char p1, char p2,
 * char p3, char p4)
 * @details Some more details about function 1
 */
```

**Comment by albert-github** [comment]
@vincentdupaquis 
What do you mean by:
> Why is there this mis-placed space, 

i.e. which misplaced space, do you mean in your code or the fact that doxygen doesn't handle it properly ( I think doxygen should handle it properly though as `char* p2` should / is equal the `char  *p2`)

**Comment by vincentdupaquis** [comment]
I mean the space in the parameter p2, it should have been 'char  *p2'. We never write 'char* p2' :)

**Comment by vincentdupaquis** [comment]
I mean `char  *p2` and `char* p2`

**Comment by doxygen** [comment]
The main problem is that `@fn` expects the function name plus arguments, but not the return type. Doxygen is normally robust if the return type is specified anyway, but the markdown processing was not yet.

**Comment by doxygen** [comment]
@vincentdupaquis Please verify if the referenced commit fixes the problem.

**Comment by albert-github** [comment]
I did some, multiline, tests as well and it looks OK to me.

**Comment by albert-github** [comment]
I think the fix here could be quite simple like, instead of
```
while (offset_<size_ && (c=data_[offset_])!=' ' && c!='\n') offset_++;
```
we use:
```
while (offset_<size_ && (c=data_[offset_])!=' ' && c!='\\' && c != '@' && c!='\n') offset_++;
```
in the `endOfLabel` function in `markdown.cpp` though I'm not 100% sure if this allowed for all usages where `endOfLabel` is used:
```
    { "a",              endOfLabel },
    { "addtogroup",     endOfLabel },
    { "anchor",         endOfLabel },
    { "b",              endOfLabel },
    { "c",              endOfLabel },
    { "cite",           endOfLabel },
    { "defgroup",       endOfLabel },
    { "e",              endOfLabel },
    { "em",             endOfLabel },
    { "emoji",          endOfLabel },
    { "enum",           endOfLabel },
    { "extends",        endOfLabel },
    { "ingroup",        endOfLabel },
    { "memberof",       endOfLabel },
    { "namespace",      endOfLabel },
    { "p",              endOfLabel },
    { "package",        endOfLabel },
    { "page",           endOfLabel },
    { "paragraph",      endOfLabel },
    { "ref",            endOfLabel },
    { "refitem",        endOfLabel },
    { "related",        endOfLabel },
    { "relatedalso",    endOfLabel },
    { "relates",        endOfLabel },
    { "relatesalso",    endOfLabel },
    { "retval",         endOfLabel },
    { "section",        endOfLabel },
    { "subpage",        endOfLabel },
    { "subsection",     endOfLabel },
    { "subsubsection",  endOfLabel },
    { "throw",          endOfLabel },
    { "throws",         endOfLabel },
    { "tparam",         endOfLabel },
    { "weakgroup",      endOfLabel },
    { "xrefitem",       endOfLabel }
```
also in
```
    { "param",          endOfParam },
```
is using the `endOfLabel` for the parameter name.

For e.g. `anchor`, `cite`, `emoji`, `*page`, `*section`, `paragrph` , `xrefitem`, `*group` I think it is for sure allowed, but for others I'm not sure as some languages might allow these special characters in identifiers.

**Comment by vincentdupaquis** [comment]
Hello,

I have tested with the project that was causing the original issue, and can confirm this fix is efficient !

Nice job :)

Best regards,
Vincent.

**Comment by albert-github** [comment]
Code has been corrected in master on GitHub (please don't close the issue as this will be done at the moment of an official release).

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.4.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
